src/goto-instrument/cover_basic_blocks.cpp

```diff
@@ -228,3 +160,46 @@ void cover_basic_blockst::update_covered_lines(block_infot &block_info)
   std::string covered_lines = format_number_range(line_list);
   block_info.source_location.set_basic_block_covered_lines(covered_lines);
 }
+
+cover_basic_blocks_javat::cover_basic_blocks_javat(
+  const goto_programt &_goto_program)
+{
+  forall_goto_program_instructions(it, _goto_program)
+  {
+    const auto &bytecode_index = it->source_location.get_java_bytecode_index();
+    if(!index_to_block.count(bytecode_index))
+    {
+      block_infos.push_back(it);
+      index_to_block.emplace(bytecode_index, block_infos.size() - 1);
+    }
+  }
+}
+
+std::size_t
+cover_basic_blocks_javat::block_of(goto_programt::const_targett t) const
+{
+  const auto &bytecode_index = t->source_location.get_java_bytecode_index();
+  const auto it = index_to_block.find(bytecode_index);
+  INVARIANT(it != index_to_block.end(), "instruction must be part of a block");
+  return it->second;
+}
+
+optionalt<goto_programt::const_targett>
+cover_basic_blocks_javat::instruction_of(const std::size_t block_nr) const
+{
+  INVARIANT(block_nr < block_infos.size(), "block number out of range");
+  return block_infos.at(block_nr);
```

Comment:
Using `at` together with the above `INVARIANT` is unnecessary. Use `block_infos[block_nr]` or drop the `INVARIANT`. The same is true below.